Firestore/core/src/firebase/firestore/util/equality.h

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EQUALITY_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EQUALITY_H_
+
+#include <memory>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * Checks if the values pointed to by two C++ pointers are equal. Two null
+ * pointers are also considered equal.
+ */
+template <typename T, typename = typename std::pointer_traits<T>::element_type>
+bool Equals(const T& lhs, const T& rhs) {
```

Comment:
I followed what is essentially the same method for Objective-C pointers, called `Equals`. I even considered moving it here.

Also, I'm mildly concerned that this isn't actually deep equality. If the type in question happens to implement that, then it is, but all this can really guarantee is one degree of depth (i.e. slightly-deep-equality).

Given the above, do you think it's worth changing `util::Equals` and `objc::Equals` both?

Reply:
This is a good objection; I don't have a strong opinion on this.

On one hand, pointers have their own built-in notion of equality, so just `Equals` doesn't immediately convey the difference in behavior (though it can be argued that the fact that the pointers are compared using a function is giving a strong hint). On the other hand, it's true that this isn't recursive deep equality.

I can try throwing a few potential names around (`PointeeEquals`? `UnrefEquals`? `ObjectEquals`?), but unless you happen to like one of them or have some other idea, I'm fine with leaving as-is, too.